libctru/source/gdbhio_dev.c

```diff
@@ -0,0 +1,196 @@
+}
+
+static ssize_t _gdbHioDevWrite(struct _reent *r, void *fdptr, const char *buf, size_t count)
+{
+	(void)r;
+	return gdbHioWrite(_gdbHioGetFdFromPtr(fdptr), buf, (unsigned int)count);
+}
+
+static off_t _gdbHioDevLseek(struct _reent *r, void *fdptr, off_t offset, int flag)
+{
+	(void)r;
+	return gdbHioLseek(_gdbHioGetFdFromPtr(fdptr), offset, flag);
+}
+
+static int _gdbHioDevRename(struct _reent *r, const char *oldpath, const char *newpath)
+{
+	(void)r;
+	return gdbHioRename(_gdbHioSkipMountpoint(oldpath), _gdbHioSkipMountpoint(newpath));
+}
+
+static int _gdbHioDevUnlink(struct _reent *r, const char *pathname)
+{
+	(void)r;
+	return gdbHioUnlink(_gdbHioSkipMountpoint(pathname));
+}
+
+static int _gdbHioDevStat(struct _reent *r, const char *pathname, struct stat *st)
+{
+	(void)r;
+	return gdbHioStat(_gdbHioSkipMountpoint(pathname), st);
+}
+
+static int _gdbHioDevFstat(struct _reent *r, void *fdptr, struct stat *st)
+{
+	(void)r;
+	return gdbHioFstat(_gdbHioGetFdFromPtr(fdptr), st);
+}
+
+static int _gdbHioDevImportFd(int gdbFd)
+{
+	int fd, dev;
+
+	dev = FindDevice("gdbhio:");
+	if(dev == -1)
+		return -1;
+
+	fd = __alloc_handle(dev);
+	if(fd == -1)
+		return -1;
+
+	*(int *)__get_handle(fd)->fileStruct = gdbFd;
+
+	return fd;
+}
+
+static const devoptab_t g_gdbHioDevoptab = {
+	.name = "gdbhio",
+	.structSize   = sizeof(int),
+	.open_r       = _gdbHioDevOpen,
+	.close_r      = _gdbHioDevClose,
+	.write_r      = _gdbHioDevWrite,
+	.read_r       = _gdbHioDevRead,
+	.seek_r       = _gdbHioDevLseek,
+	.fstat_r      = _gdbHioDevFstat,
+	.stat_r       = _gdbHioDevStat,
+	.link_r       = NULL,
+	.unlink_r     = _gdbHioDevUnlink,
+	.chdir_r      = NULL,
+	.rename_r     = _gdbHioDevRename,
+	.mkdir_r      = NULL,
+	.dirStateSize = 0,
+	.diropen_r    = NULL,
+	.dirreset_r   = NULL,
+	.dirnext_r    = NULL,
+	.dirclose_r   = NULL,
+	.statvfs_r    = NULL,
+	.ftruncate_r  = NULL,
+	.fsync_r      = NULL,
+	.deviceData   = 0,
+	.chmod_r      = NULL,
+	.fchmod_r     = NULL,
+	.rmdir_r      = NULL,
+};
+
+int gdbHioDevGettimeofday(struct timeval *tv, void *tz)
+{
+	return gdbHioGettimeofday(tv, tz);
+}
+
+int gdbHioDevIsatty(int fd)
+{
+	return gdbHioIsatty(_gdbHioGetFd(fd));
+}
+
+int gdbHioDevSystem(const char *command)
+{
+	return gdbHioSystem(command);
+}
+
+int gdbHioDevInit(void)
+{
+	int dev = FindDevice("gdbhio:");
+	if (dev >= 0) {
+		return -1;
+	}
+
+	dev = AddDevice(&g_gdbHioDevoptab);
+	return dev >= 0 ? 0 : -1;
+}
+
+void gdbHioDevExit(void)
+{
+	RemoveDevice("gdbhio:");
+}
+
+int gdbHioDevGetStdin(void)
+{
+	return _gdbHioDevImportFd(GDBHIO_STDIN_FILENO);
+}
+
+int gdbHioDevGetStdout(void)
+{
+	return _gdbHioDevImportFd(GDBHIO_STDOUT_FILENO);
+}
+
+int gdbHioDevGetStderr(void)
+{
+	return _gdbHioDevImportFd(GDBHIO_STDERR_FILENO);
+}
+
+int gdbHioDevRedirectStdStreams(bool in, bool out, bool err)
+{
+	int ret = 0;
+	if (in && (ret = dup2(gdbHioDevGetStdin(), STDIN_FILENO) < 0)) return ret;
```

Comment:
The file descriptors returned from gdbHioDevGetStd* need closed here so the user can close the new descriptors & undo the redirection without leaking resources. dup2 increases the refcount on the cloned handle so it won't be released when the first descriptor that references it is closed.